fibheap/fibheap.go

```diff
@@ -0,0 +1,384 @@
+and continue this process.  This can be shown to run in O(1) amortized time
+using yet another clever potential function.  Finally, given this function,
+we can implement delete by decreasing a key to -\infty, then calling
+dequeueMin to extract it.
+*/
+package fibheap
+
+import (
+	"fmt"
+	"math"
+)
+
+/******************************************
+ ************** INTERFACE *****************
+ ******************************************/
+
+// FloatingFibonacciHeap is an implementation of a fibonacci heap
+// with only floating-point priorities and no user data attached.
+type FloatingFibonacciHeap struct {
+	min  *Entry // The minimal element
+	size uint   // Size of the heap
+}
+
+// Entry is the entry type that will be used
+// for each node of the Fibonacci heap
+type Entry struct {
+	degree                    int
+	marked                    bool
+	next, prev, child, parent *Entry
+	// Priority is the numerical priority of the node
+	Priority float64
+}
+
+// NewFloatFibHeap creates a new, empty, Fibonacci heap object.
+func NewFloatFibHeap() FloatingFibonacciHeap { return FloatingFibonacciHeap{nil, 0} }
+
+// Enqueue adds and element to the heap
+func (heap *FloatingFibonacciHeap) Enqueue(priority float64) *Entry {
+	singleton := newEntry(priority)
+
+	// Merge singleton list with heap
+	heap.min = mergeLists(heap.min, singleton)
+	heap.size++
+	return singleton
+}
+
+// Min returns the minimum element in the heap
+func (heap *FloatingFibonacciHeap) Min() (*Entry, error) {
+	if heap.IsEmpty() {
+		return nil, fmt.Errorf("Trying to get minimum element of empty heap")
+	}
+	return heap.min, nil
+}
+
+// IsEmpty answers: is the heap empty?
+func (heap *FloatingFibonacciHeap) IsEmpty() bool {
+	return heap.size == 0
+}
+
+// Size gives the number of elements in the heap
+func (heap *FloatingFibonacciHeap) Size() uint {
+	return heap.size
+}
+
+// DequeueMin removes and returns the
+// minimal element in the heap
+func (heap *FloatingFibonacciHeap) DequeueMin() (*Entry, error) {
+	if heap.IsEmpty() {
+		return nil, fmt.Errorf("Cannot dequeue minimum of empty heap")
+	}
+
+	heap.size--
+
+	// Copy pointer. Will need it later.
+	min := heap.min
+
+	if min.next == min { // This is the only root node
+		heap.min = nil
+	} else { // There are more root nodes
+		heap.min.prev.next = heap.min.next
+		heap.min.next.prev = heap.min.prev
+		heap.min = heap.min.next // Arbitrary element of the root list
+	}
+
+	if min.child != nil {
+		// Keep track of the first visited node
+		curr := min.child
+		for ok := true; ok; ok = (curr != min.child) {
+			curr.parent = nil
+			curr = curr.next
+		}
+	}
+
+	heap.min = mergeLists(heap.min, min.child)
+
+	if heap.min == nil {
+		// If there are no entries left, we're done.
+		return min, nil
+	}
+
+	treeSlice := make([]*Entry, 0, heap.size)
+	toVisit := make([]*Entry, 0, heap.size)
+
+	for curr := heap.min; len(toVisit) == 0 || toVisit[0] != curr; curr = curr.next {
+		toVisit = append(toVisit, curr)
```

Comment:
Would it be possible to eliminate this loop and thus the toVisit allocation?  Seems like there may be some combination of checks in the outer loop immediately proceeding this one that could avoid it (although I could be wrong).